spec/UserMiddleware.coffee

```diff
@@ -0,0 +1,258 @@
+        hello: 'world'
+      receivePass passAction, action, payload, done
+      send actionIn, action, payload
+  describe 'receiving application:url action', ->
+    originalUser = null
+    beforeEach ->
+      originalUser = localStorage.getItem 'grid-user'
+      return unless originalUser
+      localStorage.removeItem 'grid-user'
+    afterEach ->
+      return unless originalUser
+      localStorage.setItem 'grid-user', originalUser
+    describe 'without logged in user', ->
+      it 'should send out an empty user:info', (done) ->
+        action = 'application:url'
+        payload = 'https://app.flowhub.io'
+        check = (data) ->
+          chai.expect(data['grid-user']).to.be.a 'null'
+        receiveAction newAction, 'user:info', check, done
+        send actionIn, action, payload
+    describe 'with logged in user', ->
+      userData =
+        id: 1
+        name: 'Henri Bergius'
+      beforeEach ->
+        localStorage.setItem 'grid-user', JSON.stringify userData
+      it 'should pass it out as-is and send user:info', (done) ->
+        action = 'application:url'
+        payload = 'https://app.flowhub.io'
+        check = (data) ->
+          chai.expect(data['grid-user']).to.eql userData
+        receiveAction newAction, 'user:info', check, done
+        send actionIn, action, payload
+    describe 'without user and with OAuth error in URL', ->
+      it 'should send the error out', (done) ->
+        action = 'application:url'
+        payload = "https://app.flowhub.io?error=redirect_uri_mismatch&error_description=The+redirect_uri+MUST+match"
+        check = (data) ->
+          chai.expect(data.message).to.contain 'The redirect_uri MUST match'
+        receiveAction newAction, 'user:error', check, done
+        send actionIn, action, payload
+    describe 'without user and with invalid grant code in URL', ->
+      xhr = null
+      code = null
+      beforeEach ->
+        code = 'dj0328hf3d9cq3c'
+        xhr = sinon.fakeServer.create()
+      afterEach ->
+        xhr.restore()
+      it 'should perform a token exchange and fail', (done) ->
+        action = 'application:url'
+        payload = "https://app.flowhub.io?code=#{code}"
+        check = (data) ->
+          chai.expect(data.message).to.contain 'bad_code_foo'
+        receiveAction newAction, 'user:error', check, done
+        send actionIn, action, payload
+        xhr.respondWith 'GET', "https://noflo-gate.herokuapp.com/authenticate/#{code}", [
+          402
+        ,
+          'Content-Type': 'application/json'
+        , JSON.stringify
+          error: 'bad_code_foo'
+        ]
+        do xhr.respond
+    describe 'without user and with grant code in URL yielding invalid API token', ->
+      xhr = null
+      code = null
+      token = null
+      beforeEach ->
+        code = 'oivwehfh24890f84h'
+        token = 'niov2o3wnnv4ioufuhh92348fh42q9'
+        xhr = sinon.fakeServer.create()
+      afterEach ->
+        xhr.restore()
```

Comment:
Would prefer to use a name for the mock which singifies testing more than `xhr`. Like `mock` or `mockAuth`